Sync/SyncTelemetry.swift

```diff
@@ -0,0 +1,120 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Shared
+import Account
+
+public enum SyncReason: String {
+    case startup = "startup"
+    case scheduled = "scheduled"
+    case backgrounded = "backgrounded"
+    case user = "user"
+    case syncNow = "syncNow"
+    case didLogin = "didLogin"
+}
+
+public protocol Stats {
+    func hasData() -> Bool
+}
+
+public struct SyncUploadStats: Stats {
+    var sent: Int = 0
+    var sentFailed: Int = 0
+
+    public func hasData() -> Bool {
+        return sent > 0 || sentFailed > 0
+    }
+}
+
+public struct SyncDownloadStats: Stats {
+    var applied: Int = 0
+    var succeeded: Int = 0
+    var failed: Int = 0
+    var newFailed: Int = 0
+    var reconciled: Int = 0
+
+    public func hasData() -> Bool {
+        return applied > 0 ||
+               succeeded > 0 ||
+               failed > 0 ||
+               newFailed > 0 ||
+               reconciled > 0
+    }
+}
+
+// TODO(sleroux): Implement various bookmark validation issues we can run into
+public struct ValidationStats: Stats {
+    public func hasData() -> Bool {
+        return false
+    }
+}
+
+public class StatsSession {
+    private var took: UInt64 = 0
+    private var startTime: Timestamp?
+
+    public func start(time: Timestamp = Date.now()) {
```

Comment:
According to the sync ping spec [1] the ping expects this to be milliseconds from epoch instead of uptime since boot which is what `uptimeNanoseconds` returns. I've pushed this using uptime but I think I should be using .now() instead...

[1] http://gecko.readthedocs.io/en/latest/toolkit/components/telemetry/telemetry/data/sync-ping.html